### PR TITLE
Implement PresenceEvent and EphemeralEvent

### PR DESCRIPTION
**Pull request by DevinR528**
Add derive for `PresenceEventContent` and create struct `AnyPreseenceEventContent` since there is only one content type.
Add derive for `EphemeralEvent` and create enum `AnyEphemeralEventContent`, convert receipt, and typing to use `derive(EphemeralEventContent)` over `ruma_api!`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -217,6 +218,15 @@ pub trait EventContent: Sized + Serialize {
     fn from_parts(event_type: &str, content: Box<RawJsonValue>) -> Result<Self, String>;
 }
 
+/// Marker trait for the content of a ephemeral event.
+pub trait EphemeralEventContent: EventContent {}
+
+/// Marker trait for the content of a presence event.
+pub trait PresenceEventContent: EventContent {}
```
Comment (jplatte) [inline src/lib.rs:225]: We don't actually need this.

File: src/presence.rs [hunk 2]
```diff
@@ -1,46 +1,56 @@
...
 }
 
-pub use ruma_common::presence::PresenceState;
+/// Informs the room of members presence.
+///
+/// This is the only event content a `PresenceEvent` can contain as it's
+/// `content` field.
+#[derive(Clone, Debug, Deserialize, Serialize, PresenceEventContent)]
+#[ruma_event(type = "m.presence")]
+pub struct AnyPresenceEventContent {
```
Comment (jplatte) [inline src/presence.rs:30]:
```suggestion
pub struct PresenceEventContent {
```

File: ruma-events-macros/src/collection.rs [hunk 3]
```diff
@@ -16,6 +16,9 @@ fn marker_traits(ident: &Ident) -> TokenStream {
             impl ::ruma_events::RoomEventContent for #ident {}
             impl ::ruma_events::MessageEventContent for #ident {}
         },
+        "AnyEphemeralEventContent" => quote! {
```
Comment (jplatte) [inline ruma-events-macros/src/collection.rs:19]:
```suggestion
        "AnyEphemeralRoomEventContent" => quote! {
```

File: src/ephemeral.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+event_content_collection! {
+    /// A ephemeral event.
+    name: AnyEphemeralEventContent,
+    events: [ "m.typing", "m.receipt" ]
+}
+
+/// Ephemeral event.
+#[derive(Clone, Debug, Event)]
+pub struct EphemeralEvent<C: EphemeralEventContent> {
```
Comment (jplatte) [inline src/ephemeral.rs:28]:
```suggestion
pub struct EphemeralRoomEvent<C: EphemeralRoomEventContent> {
```

File: src/lib.rs (hunk 1 above)
Comment (DevinR528) [reply src/lib.rs:225]: And make `PresenceEvent` non generic? I figured out a better way to do this in the macro.

Comment (jplatte) [reply src/lib.rs:225]: Yes, `PresenceEvent` should not be generic. I think I also said so before you started working on this over in the Matrix room.

Comment (DevinR528) [reply src/lib.rs:225]: Ah sorry I did see that in the Matrix room I wasn't trying to be an ass :disappointed_relieved:  I had a comment that github lost (I never submitted the review comment) about doing it with generic keeping less "cheating" branches (`if input.ident == "PresenceEvent ...`) out of the macro. So ignore all of that sorry about that.

**Comment by DevinR528** [comment]
Do you want `PresenceEvent` in it's own file?

File: src/content_enums.rs [hunk 5]
```diff
@@ -34,3 +34,9 @@ event_content_enum! {
         "m.room.topic",
     ]
 }
+
+event_content_enum! {
+    /// A ephemeral event.
```
Comment (jplatte) [inline src/content_enums.rs:39]: Missing "room", also "an" instead of "a".

File: src/lib.rs [hunk 6]
```diff
@@ -219,6 +219,9 @@ pub trait EventContent: Sized + Serialize {
     fn from_parts(event_type: &str, content: Box<RawJsonValue>) -> Result<Self, String>;
 }
 
+/// Marker trait for the content of a ephemeral event.
```
Comment (jplatte) [inline src/lib.rs:222]: 'an'

**Comment by jplatte** [comment]
It should be part of the presence module, like it is now.

**Pull request merged**